internal/getproviders/registry_client.go

```diff
@@ -292,6 +314,49 @@ func (c *registryClient) PackageMeta(provider addrs.Provider, version Version, t
 	return ret, nil
 }
 
+// findClosestProtocolCompatibleVersion searches for the provider version with the closest protocol match.
+func (c *registryClient) findClosestProtocolCompatibleVersion(provider addrs.Provider, version Version) (Version, error) {
+	var match Version
+	available, err := c.ProviderVersions(provider)
+	if err != nil {
+		return versions.Unspecified, err
+	}
+
+	// extract the maps keys so we can make a sorted list of available versions.
+	versionList := make(VersionList, 0, len(available))
+	for versionStr := range available {
+		v, err := ParseVersion(versionStr)
+		if err != nil {
+			return versions.Unspecified, ErrQueryFailed{
+				Provider: provider,
+				Wrapped:  fmt.Errorf("registry response includes invalid version string %q: %s", versionStr, err),
+			}
+		}
+		versionList = append(versionList, v)
+	}
+	versionList.Sort() // lowest precedence first, preserving order when equal precedence
+
+	protoVersions := MeetingConstraints(SupportedPluginProtocols)
+FindMatch:
+	// put the versions in increasing order of precedence
+	for index := len(available) - 1; index >= 0; index-- { // walk backwards to consider newer versions first
```

Comment:
Using `index` ranging over `len(available)` and `versionList[index]` confused me for a second. After rereading I know they're the same length, but maybe it'd be a little clearer to use `len(versionList)` in the loop?